libaleth-interpreter/VMOpt.cpp

```diff
@@ -21,22 +21,23 @@ namespace dev
 {
 namespace eth
 {
-std::array<evmc_instruction_metrics, 256> VM::c_metrics{{}};
-void VM::initMetrics()
+std::array<std::array<evmc_instruction_metrics, 256>, EVMC_MAX_REVISION> VM::s_metrics;
+std::array<std::once_flag, EVMC_MAX_REVISION> VM::s_metricsInitialized;
+
+void VM::initMetrics(evmc_revision _revision)
 {
-    static bool done = []() noexcept
-    {
-        // Copy the metrics of the top EVM revision.
-        std::memcpy(&c_metrics[0], evmc_get_instruction_metrics_table(EVMC_MAX_REVISION),
-            c_metrics.size() * sizeof(c_metrics[0]));
+    std::call_once(s_metricsInitialized[_revision], [_revision]() {
```

Comment:
`call_once` is needed to make initialization thread-safe, and we can't use the trick of static initialized by lambda as before, because now we have array of metric arrays (and we want to lazily initialize only the requested revision)

Reply:
I would do it once in https://github.com/ethereum/aleth/blob/master/libaleth-interpreter/VM.cpp#L117.

Reply:
Yeah, it would be simpler if it's for all possible revisions at one. We could also do it here as before, in static initialized by lambda call. But just wanted to avoid this unnecessary overhead.
I think I'll try that if I don't solve the GCC issue here.

Reply:
I think this is nice and generic design, but if done not lazily in `evmc_create_interpreter` we would avoid dependency on atomics and locks. I expect this to be also additional runtime dependency so you will need to link with more symbols from standard library.

I was also thinking if any of the tables are the same, but this is unlikely and we should probably not optimize for this case.